### PR TITLE
Update runner's api : all public interfaces return a "Promise" to increase usability#issue349

### PR DESCRIPTION
**Pull request by JiepengTan**
Update runner's api : all public interfaces return a "Promise" to increase usability https://github.com/goplus/spx/issues/349

## How to use: 
Please refer to :  **"cmd/spx/template/index.html"**

## Demo
project link (the zip file is too big to upload to github)：https://pan.baidu.com/s/1F38MG_Ewe8tglCnK4ChW0g?pwd=73qq 
1. download and unzip it
2. run the script (run.bat or run.sh)
```
./run.sh 8163
```

https://github.com/user-attachments/assets/1a27ca3c-e927-43fc-a94d-5b3b24531c9a

### CODE REVIEW

File: cmd/spx/template/game.js [hunk 1]
```diff
@@ -32,27 +32,92 @@ class GameApp {
...
-            addInfos = diffInfos.addInfos;
-            deleteInfos = diffInfos.deleteInfos;
-            updateInfos = diffInfos.updateInfos;
+        if (this.startProjectPromise != null) {
+            await this.startProjectPromise
+        }
+        if (this.updateProjectPromise != null) {
+            await this.updateProjectPromise
+        }
+        let promise = new Promise(async (resolve, reject) => {
```
Comment (nighca) [inline cmd/spx/template/game.js:58]: If we call `UpdateProject` multiple times:

```js
UpdateProject(...)
// While the 1st `UpdateProject` not finished, call `UpdateProject` very quickly for another 2 times
UpdateProject(...)
UpdateProject(...)
```

The 2rd & 3rd `UpdateProject` will both wait for the 1st to finish, while the 3rd will not wait for the 2rd to finish. As a result, the 2rd & 3rd `updateProject` will run "concurrently", which I'm not sure if is expected.

File: cmd/spx/template/game.js [hunk 2]
```diff
@@ -261,27 +257,25 @@ class GameApp {
...
-                this.RunGame();
+            if (!dbExists) {
+                // install project
+                // TODO store project zip and hash to indexDB 
+                this.exitFunc = this.runEditor.bind(this);
+                this.editor.init(this.basePath).then(async () => {
+                    await this.mergeProjectWithEngineRes()
+                    this.editor.copyToFS(this.tempZipPath, this.projectData);
+                    const args = ['--project-manager', '--single-window', "--install_project_name", this.projectInstallName];
+                    this.editor.start({ 'args': args, 'persistentDrops': true });
```
Comment (nighca) [inline cmd/spx/template/game.js:275]: Should we call `startProjectResolve` when `this.editor.start()` here finishes?

File: cmd/spx/template/game.js [hunk 3]
```diff
@@ -70,93 +135,24 @@ class GameApp {
         const evt = new CustomEvent('add_files', {
             detail: datas
         });
-        this.gameCanvas.dispatchEvent(evt);
+        this.editorCanvas.dispatchEvent(evt);
```
Comment (nighca) [inline cmd/spx/template/game.js:138]: Should we wait for editorCanvas to finish processing event `add_files`?

File: cmd/spx/template/game.js (hunk 2 above)
Comment (JiepengTan) [reply cmd/spx/template/game.js:275]:
```js
  await this.mergeProjectWithEngineRes()
  this.editor.copyToFS(this.tempZipPath, this.projectData);
  const args = ['--project-manager', '--single-window', "--install_project_name", this.projectInstallName];
  this.editor.start({ 'args': args, 'persistentDrops': true });
```
They are just installing the project to web IndexDB. We need to call `runEditor` to verify and import assets, so we should call `startProjectResolve` after `runEditor` finished.

```js
this.exitFunc = this.runEditor.bind(this);
```

File: cmd/spx/template/game.js (hunk 3 above)
Comment (JiepengTan) [reply cmd/spx/template/game.js:138]: Yes, this is a TODO. We should wait for `editorCanvas `to finish processing the event after calling `this.deleteFiles`. It's a bit complex and involves significant changes, so I hope to implement it in the next few commits.

File: cmd/spx/template/game.js (hunk 1 above)
Comment (JiepengTan) [reply cmd/spx/template/game.js:58]: yes, i will fixed it in the next few commits

**Review by nighca — APPROVED** [review]
issues commented will be addressed in standalone PRs.

File: cmd/spx/template/game.js (hunk 3 above)
Comment (JiepengTan) [reply cmd/spx/template/game.js:138]: this has been resove by https://github.com/goplus/spx/pull/378

File: cmd/spx/template/game.js (hunk 1 above)
Comment (JiepengTan) [reply cmd/spx/template/game.js:58]: this has been resove by https://github.com/goplus/spx/pull/378

**Pull request merged**